### PR TITLE
implement SelfSignedCertProvisioner

### LINKED ISSUE #234: Implement CertProvider Controller / Library for generating Certs for Webhooks

### PR DESCRIPTION
**Pull request by mengqiy**
Port some cert generation lib with minor change from https://github.com/kubernetes-sigs/kubebuilder/pull/203 to implement SelfSignedCertProvisioner.

Closes #234

### CODE REVIEW

**Comment by mengqiy** [comment]
GoDoc:
http://130.211.139.40/pkg/github.com/kubernetes-sigs/kubebuilder/pkg/webhook/internal/certprovisioner/

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	if err != nil {
+		return nil, errors.New("failed to generate serial number: " + err.Error())
+	}
+
+	tmpl := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               pkix.Name{Organization: []string{cp.Organization}},
+		SignatureAlgorithm:    x509.SHA256WithRSA,
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(0, 0, cp.ValidDays),
```
Comment (droot) [inline pkg/webhook/certprovisioner/selfsignedcertprovisioner.go]: If we want to unit test, then we might have to define `timeNowFn`.

**Comment by mengqiy** [comment]
Implementation has been changed to use some client-go lib. And added a test.
@droot PTAL

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"fmt"
+
+	"k8s.io/client-go/util/cert"
+)
+
+// SelfSignedCertProvisioner implements the CertProvisioner interface.
+// It provisions self-signed certificates.
+type SelfSignedCertProvisioner struct {
+	// Required Common Name
+	CommonName string
```
Comment (pwittrock) [inline pkg/webhook/certprovisioner/selfsignedcertprovisioner.go]: We need a utility function to create this name from a service `types.NamespaceName` object.

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+var _ CertProvisioner = &SelfSignedCertProvisioner{}
+
+// ProvisionServingCert creates and returns a CA certificate and certificate and
```
Comment (pwittrock) [inline pkg/webhook/certprovisioner/selfsignedcertprovisioner.go]: > ProvisionServingCert creates and returns a CA certificate and certificate and

ProvisionServingCert creates and returns a CA certificate, certificate and

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+// key for the server. serverKey and serverCert are used by the server
+// to establish trust for clients, CA certificate is used by the
+// client to verify the server authentication chain.
+// The cert will be valid for 365 days.
+func (cp *SelfSignedCertProvisioner) ProvisionServingCert() (serverKey, serverCert, caCert []byte, err error) {
```
Comment (pwittrock) [inline pkg/webhook/certprovisioner/selfsignedcertprovisioner.go]: If the return result is a struct + error then we can add new fields in the future.

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner_test.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+package certprovisioner
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"testing"
+)
+
+func TestProvisionServingCert(t *testing.T) {
+	CN := "mysvc.myns.svc"
```
Comment (pwittrock) [inline pkg/webhook/certprovisioner/selfsignedcertprovisioner_test.go]: why upper case for this variable?

**Comment by pwittrock** [comment]
lgtm.  a few comments.

**Comment by mengqiy** [comment]
PTAL

File: pkg/webhook/certprovisioner/selfsignedcertprovisioner.go (hunk 2 above)
Comment (mengqiy) [reply pkg/webhook/certprovisioner/selfsignedcertprovisioner.go]: Added

**Pull request merged**